### PR TITLE
Example VQ-VAE

### PR DESCRIPTION
**Pull request by sayakpaul**
Another example on generation. 

This example demonstrates [VQ-VAE](https://arxiv.org/abs/1711.00937). I think it's important because of the recipes it comes with. Besides, being the powerhouse for OpenAI's DALL-E, it also uses codebook sampling which is widely used in generation tasks. 

I hope the community will benefit from this. Here's the initial [Colab Notebook](https://colab.research.google.com/drive/1ixZshhVLuO8xQeF6LvTOCMFQLy3zENeF?usp=sharing).

### CODE REVIEW

**Review by fchollet — CHANGES_REQUESTED** [review]
It was a great read! Very high quality content, as usual. I pushed some copyedits.

File: examples/generative/vq_vae.py [hunk 1]
```diff
@@ -0,0 +1,568 @@
...
+        super().__init__(**kwargs)
+        self.embedding_dim = embedding_dim
+        self.num_embeddings = num_embeddings
+        self.beta = (
+            beta  # This hyperparameter is best kept between [0.25, 2] as per the paper.
+        )
+
+        # Initialize the embeddings which we will quantize.
+        w_init = tf.random_uniform_initializer()
+        self.embeddings = tf.Variable(
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply examples/generative/vq_vae.py:84]: I did this to make it a bit more explicit for `embeddings`. I feel readers not familiar with `add_weight()` method might get a bit confused in this case. When I compared two versions with and without `self.embeddings` the former felt a bit less confusing.

File: examples/generative/vq_vae.py [hunk 2]
```diff
@@ -0,0 +1,563 @@
...
+"""
+
+
+class VectorQuantizer(layers.Layer):
+    def __init__(self, num_embeddings, embedding_dim, beta=0.25, **kwargs):
+        super().__init__(**kwargs)
+        self.embedding_dim = embedding_dim
+        self.num_embeddings = num_embeddings
+        self.beta = (
+            beta  # This parameter is best kept between [0.25, 2] as per the paper.
```
Comment (fchollet) [inline examples/generative/vq_vae.py:81]: You can move the comment to the line above to avoid the parens.

**Review by fchollet — COMMENTED** [review]
Looks good! Please add the generated files.

File: examples/generative/vq_vae.py (hunk 2 above)
Comment (sayakpaul) [reply examples/generative/vq_vae.py:81]: Oh, I wanted to but it was organized by `black`.

**Comment by sayakpaul** [comment]
@fchollet added the final files. Ready for you to take over.

**Review by fchollet — APPROVED** [review]
Thank you for the great contribution!

**Pull request merged**